src/layout.rs

```diff
@@ -288,171 +277,142 @@ fn split(area: Rect, layout: &Layout) -> Rc<[Rect]> {
-
-                match *size {
-                    Constraint::Min(v) | Constraint::Max(v) => {
-                        ccs.push(elements[i].width | EQ(WEAK) | f64::from(v));
-                    }
-                    _ => {}
-                }
-            }
-        }
-        Direction::Vertical => {
-            for pair in elements.windows(2) {
-                ccs.push((pair[0].y + pair[0].height) | EQ(REQUIRED) | pair[1].y);
-            }
-            for (i, size) in layout.constraints.iter().enumerate() {
-                ccs.push(elements[i].x | EQ(REQUIRED) | f64::from(dest_area.x));
-                ccs.push(elements[i].width | EQ(REQUIRED) | f64::from(dest_area.width));
-                ccs.push(match *size {
-                    Constraint::Length(v) => elements[i].height | EQ(MEDIUM) | f64::from(v),
-                    Constraint::Percentage(v) => {
-                        elements[i].height | EQ(MEDIUM) | (f64::from(v * dest_area.height) / 100.0)
-                    }
-                    Constraint::Ratio(n, d) => {
-                        elements[i].height
-                            | EQ(MEDIUM)
-                            | (f64::from(dest_area.height) * f64::from(n) / f64::from(d))
-                    }
-                    Constraint::Min(v) => elements[i].height | GE(MEDIUM) | f64::from(v),
-                    Constraint::Max(v) => elements[i].height | LE(MEDIUM) | f64::from(v),
-                });
-
-                match *size {
-                    Constraint::Min(v) | Constraint::Max(v) => {
-                        ccs.push(elements[i].height | EQ(WEAK) | f64::from(v));
-                    }
-                    _ => {}
-                }
-            }
-        }
-    }
-    solver.add_constraints(&ccs).unwrap();
-    for &(var, value) in solver.fetch_changes() {
-        let (index, attr) = vars[&var];
-        let value = if value.is_sign_negative() {
-            0
-        } else {
-            value as u16
-        };
-        match attr {
-            0 => {
-                results[index].x = value;
-            }
-            1 => {
-                results[index].y = value;
-            }
-            2 => {
-                results[index].width = value;
-            }
-            3 => {
-                results[index].height = value;
-            }
-            _ => {}
-        }
-    }
+    {
+        use taffy::prelude::*;
+        let mut taffy = Taffy::with_capacity(layout.constraints.len());
+        let mut taffy_keys = Vec::with_capacity(layout.constraints.len());
+
+        // disabling rounding, because otherwise the following tests fail:
+        // widgets_table_columns_widths_can_use_mixed_constraints
+        // widgets_table_column_spacing_can_be_changed
+        //
+        // but disabling rounding makes the following tests fail:
+        // layout::tests::test_vertical_split_by_height
+        taffy.disable_rounding();
+
+        for elt in &layout.constraints {
```

Comment:
I think this could probably get down to something like:
```rust
let child_nodes = layout.constraints
  .iter()
  .map(|c| taffy.new_leaf((c, layout.direction).into())) // not sure what's idiomatic here
  .collect_vec()
```

Reply:
i dont think i can get it down to something like this, unless the `.map` closure is pretty long

Reply:
This compiles and runs some of the examples fine (and seems reasonably understandable even):

```rust
fn split(area: Rect, layout: &Layout) -> Rc<[Rect]> {
    use taffy::prelude::*;

    let inner = area.inner(&layout.margin);
    let mut taffy = Taffy::with_capacity(layout.constraints.len());

    // disabling rounding, because otherwise the following tests fail:
    // widgets_table_columns_widths_can_use_mixed_constraints
    // widgets_table_column_spacing_can_be_changed
    //
    // but disabling rounding makes the following tests fail:
    // layout::tests::test_vertical_split_by_height
    taffy.disable_rounding();

    let children = layout
        .constraints
        .iter()
        .map(|constraint| ConstraintAndDirection(*constraint, layout.direction).into())
        .map(|style| taffy.new_leaf(style).unwrap())
        .collect_vec();

    let root_node = taffy
        .new_with_children(
            Style {
                // setting flex_direction, because otherwise "Column" would be used for
                // "Horizontal", which makes it basically render 0 length
                flex_direction: match layout.direction {
                    Direction::Horizontal => FlexDirection::Row,
                    Direction::Vertical => FlexDirection::Column,
                },
                // root node needs to take up full space of available_space later, otherwise the
                // child elements are all length 0
                size: Size {
                    width: percent(1.0),
                    height: percent(1.0),
                },
                ..Default::default()
            },
            &children,
        )
        .unwrap();

    taffy
        .compute_layout(
            root_node,
            Size {
                width: AvailableSpace::Definite(inner.width as f32),
                height: AvailableSpace::Definite(inner.height as f32),
            },
        )
        .unwrap();

    children
        .iter()
        .map(|key| taffy.layout(*key).unwrap())
        .map(|computed| crate::layout::Rect {
            width: computed.size.width as u16,
            height: computed.size.height as u16,
            x: computed.location.x as u16 + inner.x,
            y: computed.location.y as u16 + inner.y,
        })
        .collect::<Rc<[crate::layout::Rect]>>()
}

struct ConstraintAndDirection(Constraint, Direction);
impl From<ConstraintAndDirection> for Style {
    fn from(c: ConstraintAndDirection) -> Style {
        let main_axis = match c.0 {
            Constraint::Percentage(p) => percent(p as f32 / 100.0),
            Constraint::Ratio(a, b) => percent(a as f32 / b as f32),
            Constraint::Length(length) => points(length as f32),
            Constraint::Max(max) => points(max as f32),
            Constraint::Min(min) => points(min as f32),
        };

        let (width, height) = match c.1 {
            Direction::Horizontal => (main_axis, auto()),
            Direction::Vertical => (auto(), main_axis),
        };

        match c.0 {
            Constraint::Percentage(_) | Constraint::Ratio(_, _) | Constraint::Length(_) => Style {
                size: Size { width, height },
                ..Default::default()
            },
            Constraint::Max(_) => Style {
                max_size: Size { width, height },
                ..Default::default()
            },
            Constraint::Min(_) => Style {
                min_size: Size { width, height },
                ..Default::default()
            },
        }
    }
}
```
The unit test fails in a different way however - so it's got bugs.

Reply:
Adding 
```rust
            Constraint::Max(_) => Style {
                size: auto(),
                max_size: Size { width, height },
                flex_grow: 1.0,
                ..Default::default()
            },
            Constraint::Min(_) => Style {
                size: auto(),
                min_size: Size { width, height },
                flex_grow: 1.0,
                ..Default::default()
            },
```
Gets most of the way there - the failing tests now pretty much are debatable edge cases. Especially if rounding is turned back on.

Reply:
applied in e0584f38e0b2d304bf76968aaeb4fa9c0a963539